src/link_time_code_gen.ml

```diff
@@ -8,16 +8,31 @@ type t =
   ; force_linkall : bool
   }
 
-let generate_and_compile_module cctx ~precompiled_cmi ~name:basename ~code
-      ~requires =
+let generate_and_compile_module cctx ~precompiled_cmi ~name:basename
+      ~lib ~code ~requires =
   let open Build.O in
   let sctx       = CC.super_context cctx in
   let obj_dir    = CC.obj_dir       cctx in
   let dir        = CC.dir           cctx in
   let name = Module.Name.of_string basename in
+  let wrapped = Result.ok_exn (Lib.wrapped lib) in
   let module_ =
     let src_dir = Path.build (Obj_dir.obj_dir obj_dir) in
-    Module.generated ~src_dir name
+    let gen_module = Module.generated ~src_dir name in
+    match wrapped with
+    | None -> gen_module
+    | Some (Simple false) -> gen_module
+    | Some (Simple true) ->
+      let main_module_name =
+        Lib.main_module_name lib
+        |> Result.ok_exn
+        |> Option.value_exn
+      in
+      Module.with_wrapper gen_module ~main_module_name
+    | Some (Yes_with_transition _) ->
+      Code_error.raise
+        "link time code generators for compatibiity modules is not allowed"
+        []
```

Comment:
Shouldn't we raise this eagerly when parsing the stanza? We have enough info right at this point.

Reply:
Ah yes, I thought we already had that and this case was impossible.